### PR TITLE
Adds a cone primitive 

### LINKED ISSUE #96: Add a cone primitive option for sample
**This issue is a feature request** 
A sample can be generated using primitives (Insert > Sample> Primitives).  The current options include Cuboid, Tube, Cylinder, and Sphere.

1. Add an option for a Cone primitive. This should be accessible via the menu Insert > Sample> Primitives > Cone 
2.  In the **InsertPrimitiveDialog**, The Cone parameters should be base radius (default: 100mm) and height (default: 200m)
3. Add a `create_cone` method in _primitive.py_ (sscanss.core.geometry)
4. Modify the **InsertPrimitive** class  (sscanss.app.command.insert) to create the cone mesh using method above.
5. Update docs in _add_sample.rst_
6. Update tests for cone primitve 
    - Add test for cone primitve in **testInsertPrimitiveCommand** 
    - There is no test for InsertPrimitiveDialog, add test to _test_widgets.py_

### PR DESCRIPTION
**Pull request by RabiyaF**
Closes #96 

- [x] Added an option for a Cone primitive with default radius 100 mm and default height 200 mm accessible via the menu Insert > Sample> Primitives > Cone
- [x] Updated the docs 
- [x] Updated the InsertPrimitiveCommand tests
- [x] Added the InsertPrimitiveDialog tests in test_widgets.py

### CODE REVIEW

File: sscanss/core/geometry/primitive.py [hunk 1]
```diff
@@ -6,6 +6,71 @@
...
+    :return: The vertices, normals and index array of the mesh
+    :rtype: Mesh
+    """
+
+    theta = np.linspace(0, 2 * np.pi, slices + 1)[:-1]
+
+    # Add the cone
+    r = np.repeat(np.linspace(1.0, 0.0, stacks + 1), slices)
+
+    x = np.tile(np.sin(theta), stacks + 1)
```
Comment (StephenNneji) [inline sscanss/core/geometry/primitive.py:27]: Small refactor   
```
    # Add the cone body
    r = np.repeat(np.linspace(1.0, 0.0, stacks + 1), slices) * radius

    x = np.tile(np.sin(theta), stacks + 1) * r
    y = np.tile(np.cos(theta), stacks + 1) * r

    z_div = np.linspace(0.0, 1.0, stacks + 1)
    z = np.repeat(z_div * height, slices)

    vertices = np.column_stack((x, y, z))
```

File: sscanss/core/geometry/primitive.py [hunk 2]
```diff
@@ -6,6 +6,71 @@
...
+
+    y = np.tile(np.cos(theta), stacks + 1)
+    scaled_y = np.multiply(y, r)
+
+    z_div = np.linspace(0.0, 1.0, stacks + 1)
+    z = np.repeat(z_div * height, slices)
+
+    vertices = np.column_stack((radius * scaled_x, radius * scaled_y, z))
+
+    l = np.sqrt(height**2 + radius**2)
```
Comment (StephenNneji) [inline sscanss/core/geometry/primitive.py:38]: please rename `l` to `length`

File: sscanss/core/geometry/primitive.py [hunk 3]
```diff
@@ -6,6 +6,71 @@
...
+    """Generates the vertices, normals, and indices for a cone mesh
+
+    :param radius: cone base radius
+    :type radius: float
+    :param height: cone height
+    :type height: float
+    :param slices: number of radial segments to use
+    :type slices: int
+    :return: The vertices, normals and index array of the mesh
+    :rtype: Mesh
```
Comment (StephenNneji) [inline sscanss/core/geometry/primitive.py:19]: Please update docstrings parameters

File: tests/test_widgets.py [hunk 4]
```diff
@@ -2184,6 +2183,72 @@ def testNoneSamplePropertiesDialog(self):
         self.assertEqual('0', self.dialog.sample_property_table.item(0, 1).text())
 
 
+class TestInsertPrimitiveDialog(unittest.TestCase):
```
Comment (StephenNneji) [inline tests/test_widgets.py:2186]: This only tests the default primitive arguments, you should also add tests where the values are changed and a test for validation if value is zero or any other invalid. (For these extra tests, you can test 2 primitive instead of all of them)

File: sscanss/core/geometry/primitive.py (hunk 3 above)
Comment (RabiyaF) [reply sscanss/core/geometry/primitive.py:23]: Updated

File: sscanss/core/geometry/primitive.py (hunk 2 above)
Comment (RabiyaF) [reply sscanss/core/geometry/primitive.py]: Done

File: sscanss/core/geometry/primitive.py (hunk 1 above)
Comment (RabiyaF) [reply sscanss/core/geometry/primitive.py]: Done

File: tests/test_widgets.py (hunk 4 above)
Comment (RabiyaF) [reply tests/test_widgets.py:2186]: Added additional tests

File: sscanss/core/geometry/primitive.py [hunk 5]
```diff
@@ -6,6 +6,73 @@
...
+    x = np.tile(np.sin(theta), stacks + 1) * r
+
+    y = np.tile(np.cos(theta), stacks + 1) * r
+
+    z_div = np.linspace(0.0, 1.0, stacks + 1)
+    z = np.repeat(z_div * height, slices)
+
+    vertices = np.column_stack((x, y, z))
+
+    lenght = np.sqrt(height**2 + radius**2)
```
Comment (StephenNneji) [inline sscanss/core/geometry/primitive.py:40]: typo

File: sscanss/core/geometry/primitive.py [hunk 6]
```diff
@@ -6,6 +6,73 @@
...
+
+    # Add the cone
+    r = np.repeat(np.linspace(1.0, 0.0, stacks + 1), slices) * radius
+
+    x = np.tile(np.sin(theta), stacks + 1) * r
+
+    y = np.tile(np.cos(theta), stacks + 1) * r
+
+    z_div = np.linspace(0.0, 1.0, stacks + 1)
+    z = np.repeat(z_div * height, slices)
```
Comment (StephenNneji) [inline sscanss/core/geometry/primitive.py:36]: Sorry I forgot to mention could you make it so the origin is at the centre of the cone, i.e. shift the z by `-height/2`

File: sscanss/core/geometry/primitive.py [hunk 7]
```diff
@@ -6,6 +6,73 @@
...
+    b = slices + a
+    c = np.fromiter((i if i % slices else i - slices for i in range(1, slices * stacks + 1)), int)
+    d = slices + c
+    indices = np.column_stack((a, b, c, b, d, c)).flatten()
+
+    # Add the base
+    if closed:
+        x = radius * np.sin(theta)
+        y = radius * np.cos(theta)
+        z = np.full(slices, [0])
```
Comment (StephenNneji) [inline sscanss/core/geometry/primitive.py:56]: as mention above `z = np.full(slices, [-height/2])`

File: tests/test_widgets.py [hunk 8]
```diff
@@ -2184,6 +2183,124 @@ def testNoneSamplePropertiesDialog(self):
...
+                # Testing primitive creation with custom values
+                for parameter, value in test_case['mesh_args_custom'].items():
+                    self.dialog.textboxes[parameter].value = value
+                self.dialog.create_primitive_button.click()
+                self.view.presenter.addPrimitive.assert_called_with(test_case['primitive'],
+                                                                    test_case['mesh_args_custom'])
+
+                # Testing with zero and negative values
+                for ix in range(test_case['parameter_count']):
+                    control = self.dialog.form_group.form_controls[ix]
```
Comment (StephenNneji) [inline tests/test_widgets.py:2289]: This test will always be correct if the first control (ix=0) is invalid and will never fail even if the second control is valid (ix=1)

**Review by StephenNneji — APPROVED** [review]
